### PR TITLE
[infoq] Add audio only format if available

### PR DESCRIPTION
**Pull request by wader**
Refactor cookie code into a function.
Renamed formats to http_video, http_audio, rtmp_video
Renamed extract functions to video instead of videos as they return
one or no video.

## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/rg3/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/rg3/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Add audio only format for InfoQ extractor if available

### CODE REVIEW

File: youtube_dl/extractor/infoq.py [hunk 1]
```diff
@@ -47,25 +50,43 @@ def _extract_rtmp_videos(self, webpage):
...
+    def _extract_http_audio(self, webpage):
+        http_audio_url = self._search_regex(r'<input[^>]*?name="filename"[^>]*?value="([^\"]+)"[^>]*?>', webpage, 'audio URL', fatal=False)
+        if http_audio_url is None:
+            return []
+        http_audio_url = compat_urlparse.urljoin('http://res.infoq.com/downloads/mp3downloads/', http_audio_url)
+
+        return [{
+            'format_id': 'http_audio',
+            'url': http_audio_url,
+            'ext': determine_ext(http_audio_url, ""),
```
Comment (wader) [inline youtube_dl/extractor/infoq.py]: not sure what default should be for audio only format. `determine_ext` defaults is `unknown_video`

**Comment by wader** [comment]
How do i write a test case for this? tried to find some other extractor that has multiple formats with audio only but none of them seems to test it.

I noticed that some audio URLs fail with 403 unauthorized but that seems to be the case if you try to download using a browser also.

**Comment by wader** [comment]
@yan12125 hi! do you have time and knowledge to review this?

File: youtube_dl/extractor/infoq.py [hunk 2]
```diff
@@ -47,25 +50,43 @@ def _extract_rtmp_videos(self, webpage):
...
+        return 'CloudFront-Policy=%s; CloudFront-Signature=%s; CloudFront-Key-Pair-Id=%s' % (
+            policy, signature, key_pair_id)
 
+    def _extract_http_video(self, webpage):
+        http_video_url = self._search_regex(r'P\.s\s*=\s*\'([^\']+)\'', webpage, 'video URL')
         return [{
-            'format_id': 'http',
+            'format_id': 'http_video',
             'url': http_video_url,
+            'ext': determine_ext(http_video_url),
```
Comment (yan12125) [inline youtube_dl/extractor/infoq.py]: In this case you can just omit this line as determine_ext will be automatically called under the hood.

File: youtube_dl/extractor/infoq.py [hunk 3]
```diff
@@ -47,25 +50,43 @@ def _extract_rtmp_videos(self, webpage):
...
+            'format_id': 'rtmp_video',
             'url': video_url,
             'ext': determine_ext(playpath),
             'play_path': playpath,
         }]
 
-    def _extract_http_videos(self, webpage):
-        http_video_url = self._search_regex(r'P\.s\s*=\s*\'([^\']+)\'', webpage, 'video URL')
-
+    def _extract_cookie(self, webpage):
```
Comment (yan12125) [inline youtube_dl/extractor/infoq.py]: cookie => cookies. There are 3 items.

File: youtube_dl/extractor/infoq.py [hunk 4]
```diff
@@ -47,25 +50,43 @@ def _extract_rtmp_videos(self, webpage):
...
+            'http_headers': {
+                'Cookie': self._extract_cookie(webpage)
+            },
+        }]
+
+    def _extract_http_audio(self, webpage):
+        http_audio_url = self._search_regex(r'<input[^>]*?name="filename"[^>]*?value="([^\"]+)"[^>]*?>', webpage, 'audio URL', fatal=False)
+        if http_audio_url is None:
+            return []
+        http_audio_url = compat_urlparse.urljoin('http://res.infoq.com/downloads/mp3downloads/', http_audio_url)
```
Comment (yan12125) [inline youtube_dl/extractor/infoq.py]: Is 'http://res.infoq.com/downloads/mp3downloads/' hardcoded somewhere? If so it's better to point out where it is. (for example, xyz.html or abc.js)

File: youtube_dl/extractor/infoq.py (hunk 1 above)
Comment (yan12125) [inline youtube_dl/extractor/infoq.py]: The best way is modifying determine_ext and make it recognize whether it's audio or video. As of now you can just omit this line and let the ext be unknown_video.

**Comment by yan12125** [comment]
> How do i write a test case for this? tried to find some other extractor that has multiple formats with video and audio only but none of them seems to test it.

Try to specify 'format' in 'params'. See youtube.py for some examples.

> I noticed that some audio URLs fail with 403 unauthorized but that seems to be the case if you try to download using a browser also.

You can use _is_valid_url to check URLs.

**Comment by wader** [comment]
@yan12125 Thanks for the review! Fixed most of the stuff, will have a look at tests.

I tried to use `_is_valid_url` but not sure it will work in this case. Ideally i would like to do a HEAD requests just to probe if the file is there or not.

**Comment by yan12125** [comment]
_is_valid_url connects to the specified URL and disconnect immediately. It doesn't use HEAD as some websites block HEAD requests. You can improve _is_valid_url if it doesn't fit your need.

**Comment by yan12125** [comment]
Is there an example with audio tracks?

**Comment by wader** [comment]
Ok then it should work. I will have a look at how it behaves when the response is 403.

Here is an example with audio only https://www.infoq.com/presentations/Simple-Made-Easy

File: youtube_dl/extractor/infoq.py [hunk 5]
```diff
@@ -47,25 +50,43 @@ def _extract_rtmp_videos(self, webpage):
...
+            },
+        }]
+
+    def _extract_http_audio(self, webpage):
+        http_audio_url = self._search_regex(r'<input[^>]*?name="filename"[^>]*?value="([^\"]+)"[^>]*?>', webpage, 'audio URL', fatal=False)
+        if http_audio_url is None:
+            return []
+        # base URL is found in the Location header in the response returned by
+        # GET https://www.infoq.com/mp3download.action?filename=... when logged in.
+        http_audio_url = compat_urlparse.urljoin('http://res.infoq.com/downloads/mp3downloads/', http_audio_url)
```
Comment (yan12125) [inline youtube_dl/extractor/infoq.py]: It's better to use the original ```mp3download.action``` URL here as redirected URL may change in the future.

File: youtube_dl/extractor/infoq.py [hunk 6]
```diff
@@ -47,25 +50,43 @@ def _extract_rtmp_videos(self, webpage):
...
             'url': http_video_url,
             'http_headers': {
-                'Cookie': 'CloudFront-Policy=%s; CloudFront-Signature=%s; CloudFront-Key-Pair-Id=%s' % (
-                    policy, signature, key_pair_id),
+                'Cookie': self._extract_cookies(webpage)
+            },
+        }]
+
+    def _extract_http_audio(self, webpage):
+        http_audio_url = self._search_regex(r'<input[^>]*?name="filename"[^>]*?value="([^\"]+)"[^>]*?>', webpage, 'audio URL', fatal=False)
```
Comment (yan12125) [inline youtube_dl/extractor/infoq.py]: Use _hidden_inputs or _form_hidden_inputs - they're better than the long regex pattern.

File: youtube_dl/extractor/infoq.py (hunk 5 above)
Comment (wader) [inline youtube_dl/extractor/infoq.py]: The problem is that you will only be redirected to the mp3 file if your logged in

File: youtube_dl/extractor/infoq.py (hunk 6 above)
Comment (wader) [inline youtube_dl/extractor/infoq.py]: Ok. Is there support for dom tree parsing? could'nt find anything

File: youtube_dl/extractor/infoq.py (hunk 5 above)
Comment (yan12125) [inline youtube_dl/extractor/infoq.py]: Got it. Sorry for not reading the comment carefully

File: youtube_dl/extractor/infoq.py (hunk 6 above)
Comment (yan12125) [inline youtube_dl/extractor/infoq.py]: An example is extract_attributes, which uses HTMLParser. The point here is that _hidden_inputs are much well tested. On the contrary, there are lots of tricks in HTMLParser on different Python versions.

Comment (wader) [inline youtube_dl/extractor/infoq.py]: Ok i will have a look at _hidden_inputs

**Comment by wader** [comment]
@yan12125 Think i'ved resolved all the issues now

**Comment by yan12125** [comment]
Thanks!

**Comment by wader** [comment]
@yan12125  Thanks for all help! 👍

**Pull request merged**